Add Status component to @console/shared

jtomasek opened the pull request:
* Introduces common Status components which are supposed to replace current
  `StatusIconAndText` component usage
* Expands Icon + Text implementation with Link and Popover capabilities
* Allows to define custom resource specific status handling while reusing existing status components (important for plugins)
* Allows to easily define custom status components while reusing underlining components provided
* Current `StatusIconAndText` usage is replaces by `Status` component
* Uses PatternFly4 icons

Next steps:
- [x] Replace current usage of original `StatusIconAndText` component with `Status` component
- [x] remove original `StatusIconAndText` component

christianvogt commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,48 @@
+
+export const Status = ({ icon, children }: StatusProps) => (
+  <>
+    {icon && <Icon type="pf" name={icon} className="kubevirt-status__icon" />} {children}
```
Should the space use `{' '}` to ensure it is rendered? But then you need to do more logic to only include the space of the icon is rendered.

Not exactly sure if the purpose but is it supposed to allow any icon as status?

jtomasek replied:
Originally the space was not included and margin styling was used to ensure the space is present. I'd prefer using space as it adjusts to the font size. I'll fix this to include the space only when icon is present.

christianvogt commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,48 @@
+
+export const PopoverStatus = ({ icon, header, children }: PopoverStatusProps) => (
+  <Popover position="right" headerContent={header} bodyContent={children}>
+    <span className="kubevirt-status__popover">
```
We should not use non console classnames in shared components.

jtomasek replied:
Yep, I forgot to remove/change this.

christianvogt commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,48 @@
+  <Popover position="right" headerContent={header} bodyContent={children}>
+    <span className="kubevirt-status__popover">
+      <Status icon={icon}>
+        <Button className="kubevirt-status__button" bsStyle="link">
```
A button without any event handling?

jtomasek replied:
The Button is internally enhanced by Popover and acts as a toggle (https://patternfly-react.surge.sh/patternfly-4/components/popover/)

christianvogt replied:
oh pf...

christianvogt commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,48 @@
+);
+
+type LinkStatusProps = StatusProps & {
+  linkMessage?: string;
```
Rename to linkTitle to be consistent with usage

christianvogt commented on frontend/packages/console-shared/src/components/status/__tests__/status.test.tsx:
```diff
@@ -0,0 +1,14 @@
+describe('<Status />', () => {
+  it('renders correctly', () => {
+    const component = shallow(testStatus());
+    expect(component).toMatchSnapshot();
```
The snapshot seems to have a lot of noise.
I also see little value in this test case.

vojtechszocs replied:
I think this relates to the common practice where each React component has (at least) a simple snapshot test to ensure the rendered markup doesn't change unexpectedly.

As for the messy snapshot file, there are things which don't make sense to serialize (like symbols), meaning we should update Jest & Enzyme & related dependency versions and ensure the generated snapshots are meaningful.

Console doesn't currently have any snapshot tests (`git grep 'toMatchSnapshot'` returns nothing) so we should revisit the Enzyme (to-JSON) serializer stuff here. This also relates to #1648 which I'm revisiting at the moment.

jtomasek commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,48 @@
+  children: React.ReactNode;
+};
+
+export const Status = ({ icon, children }: StatusProps) => (
```
When I use React.FC, current linting setup forces me to explicitly add type for props which introduces unnecessary duplication:
```
export const Status: React.FC<StatusProps> = ({ icon, children }: StatusProps) => (...
```

christianvogt replied:
That's not right. I am currently updating dev console in the monorepo and do not encounter this. Could you provide the error please and indicate what typescript version you are using (hopefully the one from the workspace).

jtomasek replied:
This is the error I am getting: https://github.com/openshift/console/pull/1539#pullrequestreview-245356378

vojtechszocs replied:
IIUC this is addressed by #1721 (see [ESLint update commit](https://github.com/openshift/console/pull/1721/commits/9264baa99d4cbedb3e0e3ae2bee2e5f2d8dfbe47)).

christianvogt commented:
@spadgett @vojtechszocs 
How do you want to treat the shared package. Should we define some criteria?
Anything generic or anything that is used across multiple packages?

christianvogt commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,48 @@
+
+export const Status = ({ icon, children }: StatusProps) => (
+  <>
+    {icon && <Icon type="pf" name={icon} className="kubevirt-status__icon" />} {children}
```
Is this suppose to be limited to certain "status" icons? or can they be anything?
PF4 icons don't work this way as they are component based.

vojtechszocs replied:
`Icon` comes from `patternfly-react` package which means PF3 ([component showcase](https://patternfly-react.surge.sh/patternfly-3/index.html)).

Supported `type` includes `pf` (PatternFly specific) and `fa` (FontAwesome complementary). `name` depends on `type`.

In this case, since `type` is hardcoded to `pf`, `name` could be checked against a list of available PF icon names, but I'm not sure if such list exists.

vojtechszocs commented:
> @spadgett @vojtechszocs
> How do you want to treat the shared package. Should we define some criteria?
> Anything generic or anything that is used across multiple packages?

Good question. I see that `console-shared` has constants and selectors used currently by `metal3-plugin` only. I'd expect other plugin packages to eventually use these as well.

My understanding is that `console-shared` should contain "anything that is used across multiple packages". (KubeVirt, Metal3 and other plugins are still in their early stages and their code will grow and improve over time.)

vojtechszocs commented on frontend/packages/console-shared/src/components/status/fixtures/status.fixture.tsx:
```diff
@@ -0,0 +1,23 @@
+
+import { Status, PopoverStatus } from '../status';
+
+export default [
```
@jtomasek If I understand correctly, this is the [Cosmos](https://github.com/react-cosmos/react-cosmos) fixture format, right?

@christianvogt This could be another use case for `console-scripts` package - add Cosmos dependency & configuration and use it in packages that wish to leverage this tool.

jtomasek commented:
/test e2e-aws-console-olm

spadgett commented:
> Next step: Replace current usage of original `StatusIconAndText` component with `GenericStatus` component, remove original `StatusIconAndText` component

This seems like the wrong way to approach this. There are two parts here:

1. Moving the status icon components into `@console/shared`.
2. Updating them with new capabilities.

This looks like you copied status-icon.tsx and made some changes. The code will drift over time if we do this, and it might be hard to get rid of the old component. We're still trying to remove list.tsx as an example.

I think (1) and (2) should be separate PRs, and we shouldn't introduce duplicate components.

/hold

spadgett commented:
> * Uses PatternFly4 icons

What icons are you changing? It's hard to tell because you've copied the code to another file, so there's no diff :) 

We just recently updated all the icons for this design:

https://openshift.github.io/openshift-origin-design/web-console/4.0-designs/status/status

There were some specific reasons we chose the ones we did.

jtomasek commented:
> > * Uses PatternFly4 icons
> 
> What icons are you changing? It's hard to tell because you've copied the code to another file, so there's no diff :)

I am not changing any icons, I just used PF4 component based counterparts. https://patternfly-react.surge.sh/patternfly-4/icons/Icons/

christianvogt commented on frontend/packages/console-shared/src/components/status/generic-status.tsx:
```diff
@@ -0,0 +1,77 @@
+  status?: string;
+};
+
+export const GenericStatus: React.FC<{ status?: string }> = ({ status, children }) => {
```
I think `GenericStatus` is a bad name.
`React.FC<{ status?: string }>` should reference the prop type

christianvogt commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,88 @@
+    <StatusIconAndText icon={icon} title={title} />
+  );
+
+export const SuccessStatus: React.FC<{ title: string }> = ({ title }) => (
```
Use a pre-defined type for the react props.

christianvogt commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,88 @@
+};
+
+export const StatusIconAndText: React.FC<StatusProps> = ({ icon, title, spin }) => {
+  if (!title) {
```
If there is an icon but no title, you don't want to show the icon?

jtomasek replied:
I don't think we want to show icon when there is no title. I am sticking with the original StatusIconAndText logic: https://github.com/openshift/console/blob/master/frontend/public/components/utils/status-icon.tsx#L72-L74

christianvogt replied:
ok

christianvogt commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,88 @@
+import { DASH } from '../../constants';
+
+type StatusProps = {
+  icon?: React.ReactElement;
```
Is the intention to only support patternfly react icons? Or any component? If it's only a patternfly react icon, we can change the prop types to accept `color` and `size` just like the pattern fly react icons, and instead of `ReactElement` accept a component type. Then you can render it without the need to `cloneElement`.

christianvogt commented:
@spadgett (fyi @vojtechszocs) for shared components (and any new code in packages in general) do we want to adopt PascalCase component naming along with expecting a single default component export and multiple files, over single large file with multiple components? It's a fairly wide adopted convention and one that patternfly also follows.

jtomasek commented:
> do we want to adopt PascalCase component naming along with expecting a single default component export and multiple files, over single large file with multiple components?
+1

andybraren commented:
> - Expands Icon + Text implementation with Link and Popover capabilities
> - Allows to define custom resource specific status handling while reusing existing status components (important for plugins)

Just `+1`'ing these from a UX perspective. Status Popovers, [documented here in the design repo](http://openshift.github.io/openshift-origin-design/web-console/4.0-designs/status/status#status-popovers), are a particularly important part of the UX for the new objects introduced by KubeVirt, Metal3, and some Storage objects as well.

It sounds like the technical approach here needs some tweaking, but UXD agrees that more flexibility around how Statuses can be labeled and interacted with (in both Tables and Resource Details pages) would be a great improvement.

christianvogt commented on frontend/public/components/operator-lifecycle-manager/descriptors/status/index.tsx:
```diff
@@ -79,7 +80,7 @@ export const StatusDescriptor: React.SFC<DescriptorProps> = (props) => {
     <dd className="olm-descriptor__value">
       {descriptor.displayName === 'Status' ?
-        (<StatusIconAndText status={value} iconName={value === 'Running' ? 'ok' : undefined} />) :
+        ({value === 'Running' ? <SuccessStatus status={value} /> : <Status status={value} />) :
```
Just some thoughts.

I'd like to be able to use the same `Status` component here instead. But we should then enumerate on a set of accepted status values in the prop types:
`<Status status={value === 'Running' && 'Complete' || value} />`

Alternatively, provide bool props (for the pre-defined finite set of concrete statuses) that take precedence over the status:
`<Status status={value} success={value === 'Running'} />`

jtomasek replied:
> Just some thoughts.
> 
> I'd like to be able to use the same `Status` component here instead. But we should then enumerate on a set of accepted status values in the prop types:
> `<Status status={value === 'Running' && 'Complete' || value} />`

I actually initially did this initially when I created this patch. But then I realised it is not possible as the status passed as prop is used as a `title` for the StatusIconAndText so the result in your example when `value === 'Running'` results in `Icon + 'Complete'` instead of `Icon + 'Running'`

> 
> Alternatively, provide bool props (for the pre-defined finite set of concrete statuses) that take precedence over the status:
> `<Status status={value} success={value === 'Running'} />`

I feel like expanding Status component to address this may be a bit overkill. What we need is to provide a way to define mapping of a certain (custom) status string to predefined status component (which ensures to render correct icon)

e.g.:
```
switch (status) {
  case 'discovered':
    return <SuccessStatus title={status} />;
  case 'discovering':
    return <ProgressStatus title={status} />;
  case 'discovery error':
    return <ErrorStatus title={status}>{errorMessage}</ErrorStatus>;
  default:
    return <Status status={status} /> // fall back to Status component for other states
}
```

This IMHO nicely says: "`discovered` is a `success` status"...

Comparing this to current solution which says: "Make 'Running' status act the same as 'Complete' status"

There is currently also one other way to define custom status existing implementation and that is by using `iconName` on the StatusIconAndText component. (see https://github.com/openshift/console/pull/1774/files#diff-2b92d7e552b962318adc922d3e6f4b61R39). I find this also quite bad as it forces you to specify icon name, which can lead to icon usage inconsistency and problems when icons are refactored/changed.

@spadgett noted a problem of ensuring to achieve consistency of same icon always being used for given status string. I think my changes don't make the current situation worse in this regard. Your example shows, there are cases when it is necessary to deviate from this. My changes add standardised status components which should help avoid this.

christianvogt replied:
heh of course i wasn't thinking about title being the status itself ... argh

vojtechszocs commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,77 @@
+};
+
+export const Status: React.FC<StatusProps> = ({ status, children }) => {
+  switch (status) {
```
It seems that the concepts of `status` and `title` are merged together at the moment, carried over from the existing `StatusIconAndText` component design.

With this `Status` component, we could map states (statuses?) to titles if we ever decide to decouple these concepts. In general, the `status` string should be the key indicator to determine icon and any other data needed to render the underlying `StatusIconAndText`.

jtomasek replied:
Done

vojtechszocs commented on frontend/packages/console-shared/src/components/status/status.tsx:
```diff
@@ -0,0 +1,77 @@
+    case 'Completed':
+    case 'Enabled':
+    case 'Ready':
+    case 'Up to date':
```
Nit: this is inconsistent with other status values where words are always capitalized, but I guess we want to keep it this way for now.

vojtechszocs commented on frontend/packages/console-shared/src/components/status/status-components.tsx:
```diff
@@ -0,0 +1,100 @@
+  return (
+    <>
+      {icon &&
+        React.cloneElement(icon, {
```
Nit: I think we should avoid `React.cloneElement` if possible.

Did you consider using an `Icon` prop that references a React component type, instead of its instance?

For example: `Icon && <Icon className={...} otherProp={...} />`

jtomasek replied:
@christianvogt suggested the same, that solution prevents ability to pass custom props to the icon. We would have to introduce additional iconProps prop to satisfy that. Maybe passing additional props is YAGNI but I would rather keep that option available to provide a way to pass in custom className or id which IMHO can be a valid use case.

vojtechszocs commented:
> > do we want to adopt PascalCase component naming along with expecting a single default component export and multiple files, over single large file with multiple components?
> 
> +1

+1 from me as well.

vojtechszocs commented:
> adopt PascalCase component naming along with expecting a single default component export and multiple files

If we could use ESLint to check this (in a way that fits the modular `eslint-plugin-console` "build your own config out of X presets" approach), that'd be great. Any package could opt-into that convention and enforce it via the linter workflow.

spadgett commented:
> Just `+1`'ing these from a UX perspective. Status Popovers, [documented here in the design repo](http://openshift.github.io/openshift-origin-design/web-console/4.0-designs/status/status#status-popovers), are a particularly important part of the UX for the new objects introduced by KubeVirt, Metal3, and some Storage objects as well.

But we should look to add this sort of popover consistently throughout console instead of just these specific plugins. My concern is that we're only adding it in a few places and not looking more broadly at existing console pages. Notably we should align the project workloads tab error reporting with these status popovers. They have a similar purpose, but are implemented differently (tooltips). But we also show status in many other places (pods, builds, routes, etc.).

christianvogt commented on frontend/packages/console-shared/src/components/status/status-components.tsx:
```diff
@@ -0,0 +1,100 @@
+
+export const StatusIconAndText: React.FC<StatusIconAndTextProps> = ({ icon, title, spin }) => {
+  if (!title) {
+    return <>{DASH}</>;
```
Unfortunately not! Typescript bug ;/

cshinn commented:
> > Just `+1`'ing these from a UX perspective. Status Popovers, [documented here in the design repo](http://openshift.github.io/openshift-origin-design/web-console/4.0-designs/status/status#status-popovers), are a particularly important part of the UX for the new objects introduced by KubeVirt, Metal3, and some Storage objects as well.
> 
> But we should look to add this sort of popover consistently throughout console instead of just these specific plugins. My concern is that we're only adding it in a few places and not looking more broadly at existing console pages. Notably we should align the project workloads tab error reporting with these status popovers. They have a similar purpose, but are implemented differently (tooltips). But we also show status in many other places (pods, builds, routes, etc.).

@spadgett Is it worth making a push to quickly find other places in the UI where this info would be helpful so that the addition isn't too jarring? If we were able to provide locations and content, how hard would it be to add to more pages?

jtomasek commented:
> @spadgett Is it worth making a push to quickly find other places in the UI where this info would be helpful so that the addition isn't too jarring? If we were able to provide locations and content, how hard would it be to add to more pages?

As part of this PR I am replacing all occurrences of existing StatusIconAndText and replacing it with new Status component (second commit https://github.com/openshift/console/pull/1697/commits/099db40edddcf4c08dfbdae258a41e7b078ef60d). I think we can easily go through these replacements and identify if there is an error message available and add the message to <Status> component as children, so it gets automatically included in popover.

I found out that there are 2 places where `StatusIcon` is used as a standalone, so I'll update this PR to allow to cover this use case as well. Then we can completely remove status-icon.tsx to make the transition complete.

jtomasek commented:
- added optional `title` prop to `Status` component to allow specify title explicitly, if not provided, status string is used
- added optional `iconOnly` prop to allow rendering status icon only without text
- refactored status components into separate modules, PascalCase names, default exports
- replaced all usage of original `StatusIconAndText` and `StatusIcon` components
- removed status-icon.tsx

jtomasek commented:
/retest

jtomasek commented:
/test frontend

jtomasek commented:
rebased

jtomasek commented:
rebased

jtomasek commented:
/retest

jtomasek commented:
/retest

jtomasek commented:
/retest

vojtechszocs commented on frontend/packages/console-shared/src/components/status/Status.tsx:
```diff
@@ -0,0 +1,90 @@
+  const t = title || status;
+  switch (status) {
+    case 'New':
+      return <StatusIconAndText icon={<HourglassStartIcon />} title={t} iconOnly={iconOnly} />;
```
Nit: we can represent the commonly occuring props as

```tsx
const statusProps = { title: title || status, iconOnly };
// ...
<StatusIconAndText {...statusProps} icon={<HourglassStartIcon />} />;
```

The `{...statusProps}` spread should come first, so that we can still customize e.g. `title` or `status` if necessary.

jtomasek replied:
Nice!

vojtechszocs commented on frontend/packages/console-shared/src/components/status/StatusIconAndText.tsx:
```diff
@@ -0,0 +1,34 @@
+};
+
+const StatusIconAndText: React.FC<StatusIconAndTextProps> = ({ icon, title, spin, iconOnly }) => {
+  if (!title) {
```
Just wondering, if `title` isn't defined but `icon` is, we'll still render a dash - is this an expected behavior?

jtomasek replied:
Probably not, I'll revisit.

jtomasek replied:
Yes, this is intentional as it allows to handle the case when `StatusIconAndText` component is used and the status (`title` prop) is not yet set. There is explicit `iconOnly` prop to render an icon without text.

vojtechszocs replied:
OK, thanks for clarification.

vojtechszocs commented on frontend/public/components/operator-lifecycle-manager/descriptors/status/index.tsx:
```diff
@@ -79,9 +80,9 @@ export const StatusDescriptor: React.SFC<DescriptorProps> = (props) => {
-        descriptor.displayName === 'Status'
-          ? (<StatusIconAndText status={value} iconName={value === 'Running' ? 'ok' : undefined} />)
-          : (<Capability descriptor={descriptor} capability={capability} value={value} namespace={namespace} />)
+        descriptor.displayName === 'Status' ?
```
Nit: I think the convention is to put `?` and `:` at the beginning of their respective expressions

```tsx
descriptor.displayName === 'Status'
  ? (value === 'Running' ? <SuccessStatus title={value} /> : <Status status={value} />)
  : <Capability descriptor={descriptor} capability={capability} value={value} namespace={namespace} />
```

jtomasek replied:
Done

vojtechszocs commented on frontend/public/components/overview/build-overview.tsx:
```diff
@@ -50,17 +51,23 @@ const BuildOverviewItem: React.SFC<BuildOverviewListItemProps> = ({build}) => {
+  const statusTitle = (
+    <>
+      Build
+      &nbsp;
```
Nit: I think we should prefer `{' '}` instead of `&nbsp;` HTML character entity.

jtomasek replied:
I think the &nbsp; is used here on purpose to prevent line breaks (I did not introduce this as part of my change)

vojtechszocs replied:
OK, understood.

vojtechszocs commented:
LGTM overall with some minor comments.

jtomasek commented:
Addressed the comments and rebased.

vojtechszocs commented on frontend/public/components/build.tsx:
```diff
@@ -3,6 +3,7 @@ import * as _ from 'lodash-es';
 import { Link } from 'react-router-dom';
 import * as classNames from 'classnames';
 import { sortable } from '@patternfly/react-table';
+import { Status } from '@console/shared';
```
Nit: newline after vendor imports.

vojtechszocs commented on frontend/public/components/persistent-volume-claim.jsx:
```diff
@@ -2,11 +2,12 @@ import * as React from 'react';
 import * as _ from 'lodash-es';
 import { sortable } from '@patternfly/react-table';
 import * as classNames from 'classnames';
+import { Status } from '@console/shared';
```
Nit: newline after vendor imports.

vojtechszocs commented on frontend/public/components/persistent-volume.jsx:
```diff
@@ -2,13 +2,14 @@ import * as React from 'react';
 import * as _ from 'lodash-es';
 import { sortable } from '@patternfly/react-table';
 import * as classNames from 'classnames';
+import { Status } from '@console/shared';
```
Nit: newline after vendor imports.

vojtechszocs commented on frontend/public/components/routes.tsx:
```diff
@@ -4,8 +4,9 @@ import * as classNames from 'classnames';
 import { Popover } from '@patternfly/react-core';
 import { sortable } from '@patternfly/react-table';
 import { QuestionCircleIcon } from '@patternfly/react-icons';
+import { Status } from '@console/shared';
```
Nit: newline after vendor imports.

vojtechszocs commented on frontend/public/components/template-instance.tsx:
```diff
@@ -2,6 +2,7 @@ import * as React from 'react';
 import * as _ from 'lodash-es';
 import * as classNames from 'classnames';
 import { sortable } from '@patternfly/react-table';
+import { Status } from '@console/shared';
```
Nit: newline after vendor imports.

vojtechszocs commented on frontend/public/components/job.jsx:
```diff
@@ -2,10 +2,11 @@ import * as React from 'react';
 import { Link } from 'react-router-dom';
 import * as classNames from 'classnames';
 import { sortable } from '@patternfly/react-table';
+import { Status } from '@console/shared';
```
Nit: newline after vendor imports.

vojtechszocs commented on frontend/public/components/node.tsx:
```diff
@@ -2,12 +2,13 @@ import * as _ from 'lodash-es';
 import * as React from 'react';
 import * as classNames from 'classnames';
 import { sortable } from '@patternfly/react-table';
+import { Status } from '@console/shared';
```
Nit: newline after vendor imports.

vojtechszocs commented on frontend/public/components/operator-lifecycle-manager/operand.tsx:
```diff
@@ -4,6 +4,7 @@ import * as _ from 'lodash-es';
 import { connect } from 'react-redux';
 import * as classNames from 'classnames';
 import { sortable } from '@patternfly/react-table';
+import { Status, SuccessStatus } from '@console/shared';
```
Nit: newline after vendor imports.

jtomasek commented:
Rebased

jtomasek commented:
/test e2e-aws

vojtechszocs commented:
/lgtm

spadgett commented:
/hold cancel

jelkosz commented:
/test e2e-aws-console

jtomasek commented:
/test e2e-aws-console

jtomasek commented:
Rebased, fixed conflict in `packages/metal3-plugin/src/components/host-status.tsx`

jtomasek commented:
/test e2e-aws-console-olm
/test e2e-aws

honza commented:
/lgtm
